### PR TITLE
Emit warnings on unused parens in index expressions

### LINKED ISSUE #96606: [Diagnostic ER] Missing unnecessary parentheses warning
rustc could give a unnecessary parentheses warning here:

```rust
fn main() {
    let arr = [0; 10];
    let i = 2;
    let y = arr[(5 - i)];
    dbg!(y);
}
```

### PR DESCRIPTION
**Pull request by PossiblyAShrub**
Fixes: #96606.

I am not sure what the best term for "index expression" is. Is there a better term we could use?

### CODE REVIEW

**Review by lcnr — COMMENTED** [review]
can you also add a text for unused braces in index expressions?

Using "index expression" in the diagnostics seems good to me as that terminology is also used in the reference.

after adding that test, r=me :+1:

**Comment by PossiblyAShrub** [comment]
> can you also add a text for unused braces in index expressions?

Done.

@rustbot ready

File: tests/ui/lint/unused/issue-96606.rs [hunk 1]
```diff
@@ -0,0 +1,7 @@
+#[deny(unused)]
+fn main() {
+    let arr = [0; 10];
+    let _ = arr[(0)]; //~ ERROR unnecessary parentheses around index expression
+    let _ = arr[{0}]; //~ ERROR unnecessary braces around index expression
```
Comment (lcnr) [inline tests/ui/lint/unused/issue-96606.rs:5]:
```suggestion
    let _ = arr[{0}]; //~ ERROR unnecessary braces around index expression
    let _ = arr[{ let x = 0; x }];
```
here we shouldn't lint on the braces, please also add that as a test.

**Comment by lcnr** [comment]
after the last nit, you can approve this pr yourself using `@bors r=lcnr rollup`.

@bors delegate+

**Comment by PossiblyAShrub** [comment]
Added the extra test. Thanks for the review!

@bors r=lcnr rollup

**Comment by PossiblyAShrub** [comment]
Hmmm. Looks like bors doesn't want to let merge this. I'll just mark it as S-waiting-on-review. (Let me know if I should do something different as I am still new to the contribution process.)

@rustbot ready

**Comment by lcnr** [comment]
weird, not sure why the `delegate` was ignored :sweat_smile: 

@bors r+ rollup

thanks

**Pull request merged**